File: LISTA 1/BatalhaNaval.java
//Nome: Pedro Elias Santiago Mattar

import java.util.Scanner;

public class BatalhaNaval {
    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        int[][] coordenadasJogador1 = new int[5][2], coordenadasJogador2 = new int[5][2];
        int vencedor = 0;
        char[][] mapaJogador1 = new char[5][5], mapaJogador2 = new char[5][5];
        boolean turnoJogador1 = true;

        // gera os mapas de cada jogador
        geraMapas(mapaJogador1);
        geraMapas(mapaJogador2);

        System.out.println("Welcome to Battleship!\n\n");

        // Solicita e marca as coordenadas dos navios do jogador 1 no mapa
        solicitaCoordenadas(1, coordenadasJogador1);
        marcaNaviosNoMapa(mapaJogador1, coordenadasJogador1);
        imprimeMapa(mapaJogador1);

        // Solicita e marca as coordenadas dos navios do jogador 2 no mapa
        solicitaCoordenadas(2, coordenadasJogador2);
        marcaNaviosNoMapa(mapaJogador2, coordenadasJogador2);
        imprimeMapa(mapaJogador2);

        // Gera os mapas novamente, para ocultar a posição dos navios dos jogadores
        geraMapas(mapaJogador1);
        geraMapas(mapaJogador2);

        // Loop principal do jogo
        while (vencedor == 0) {
            if (turnoJogador1) {
                vencedor = turnoJogador(1, mapaJogador2, coordenadasJogador2, scanner) ? 1 : 0;
            } else {
                vencedor = turnoJogador(2, mapaJogador1, coordenadasJogador1, scanner) ? 2 : 0;
            }
            turnoJogador1 = !turnoJogador1;
        }

        if (vencedor > 0) {
            System.out.println("PLAYER " + vencedor + " WINS! YOU SUNK ALL OF YOUR OPPONENT’S SHIPS!");
            System.out.println("Final boards:");
            System.out.println("Jogador 1:");
            imprimeMapa(mapaJogador1);
            System.out.println("Jogador 2:");
            imprimeMapa(mapaJogador2);
        }
    }

    // função que executa a lógica de um turno de um jogador
    public static boolean turnoJogador(int jogador, char[][] mapaJogador, int[][] coordenadasJogador, Scanner scanner) {
        BatalhaNaval jogo = new BatalhaNaval();
        boolean entradaValida = false;
        boolean localJaAtacado = false;
        int x, y;

        while (!entradaValida || localJaAtacado) {
            System.out.printf("Player %d, enter hit row/column: ", jogador);
            x = scanner.nextInt();
            y = scanner.nextInt();
            entradaValida = validaCoordenadas(x, y); // Verifica se as coordenadas são válidas.
            if (entradaValida) {
                localJaAtacado = verificaLocaisAtacados(x, y, mapaJogador); // Verifica se o local já foi atacado.
                if (localJaAtacado) {
                    System.out.println("You already fired on this spot. Choose different coordinates.");
                } else {
                    marcaLocaisMapa(x, y, mapaJogador, coordenadasJogador, jogador, jogador == 1 ? 2 : 1);
                    imprimeMapa(mapaJogador);
                }
            }
        }

        return jogo.todosNaviosAfundados(mapaJogador, jogador) == 5;
    }

    // função que verifica se houve um vencedor
    public static int verificaVencedor(char[][] mapaJogador1, char[][] mapaJogador2) {
        BatalhaNaval jogo = new BatalhaNaval();
        if (jogo.todosNaviosAfundados(mapaJogador2, 1) == 5) {
            return 1;
        }
        if (jogo.todosNaviosAfundados(mapaJogador1, 2) == 5) {
            return 2;
        }
        return 0;
    }

    // solicita que o jogador informe as coordenadas nas quais os navios serão
    // colocados
    private static void solicitaCoordenadas(int jogador, int[][] coordenadas) {
        Scanner scanner = new Scanner(System.in);
        boolean entradaValida;
        boolean coordenadaAnteriorValida;

        System.out.println("PLAYER " + jogador + ", ENTER YOUR SHIPS COORDINATES.");

        for (int i = 0; i < 5; i++) {
            entradaValida = false;
            coordenadaAnteriorValida = false;

            // Laço para verificar a validade das coordenadas informadas
            while (!entradaValida || !coordenadaAnteriorValida) {

                System.out.println("Enter ship " + (i + 1) + " location:");

                coordenadas[i][0] = scanner.nextInt();
                coordenadas[i][1] = scanner.nextInt();

                // Verifica se as coordenadas informadas são válidas
                entradaValida = validaCoordenadas(
                        coordenadas[i][0],
                        coordenadas[i][1]);

                if (i > 0) {
                    coordenadaAnteriorValida = validaCoordenadasAnteriores(
                            coordenadas[i][0],
                            coordenadas[i][1],
                            coordenadas,
                            i);
                } else {
                    coordenadaAnteriorValida = true;
                }
            }
        }
    }

    // verifica se todos os navios já foram afundados
    private int todosNaviosAfundados(char[][] mapa, int jogador) {
        int cont = 0;
        for (int i = 0; i < mapa.length; i++) {
            for (int j = 0; j < mapa[0].length; j++) {
                if (mapa[i][j] == 'X') {
                    cont++;
                }
            }
        }

        return cont;
    }

    // verifica se as coordenadas correspondentes já foram atacadas
    public static boolean verificaLocaisAtacados(int x, int y, char[][] mapa) {
        if (mapa[x][y] == 'X' || mapa[x][y] == 'O') {
            return true;
        }

        return false;
    }

    // marca locais de ataque (X) ou locais de erro (O) no mapa do jogador que está
    // sendo atacado
    public static void marcaLocaisMapa(int x, int y, char[][] mapa, int[][] coordenadas, int jogadorAtaque,
            int jogadorDefesa) {
        boolean acertou = false;
        for (int i = 0; i < coordenadas.length; i++) {
            if (x == coordenadas[i][0] && y == coordenadas[i][1]) {
                mapa[x][y] = 'X';
                System.out.println("PLAYER " + jogadorAtaque + " HIT PLAYER " + jogadorDefesa + "'s SHIP!");
                acertou = true;
                break;
            }
        }
        if (!acertou && mapa[x][y] == '-') {
            mapa[x][y] = 'O';
            System.out.println("PLAYER " + jogadorAtaque + " MISSED!");
        }
    }

    // marca o local dos navios no mapa
    public static void marcaNaviosNoMapa(char[][] mapa, int[][] coordenadas) {
        for (int i = 0; i < coordenadas.length; i++) {
            int x = coordenadas[i][0];
            int y = coordenadas[i][1];
            mapa[x][y] = '@';
        }
    }

    // verifica se as coordenadas informadas pelo usuário já não foram informadas
    // previamente
    public static boolean validaCoordenadasAnteriores(int x, int y, int[][] coordenadas, int numNavios) {
        for (int i = 0; i < numNavios; i++) {
            if (x == coordenadas[i][0] && y == coordenadas[i][1]) {
                System.out.println("You already have a ship there. Choose different coordinates.");
                return false;
            }
        }

        return true;
    }

    // imprime o mapa do jogador selecionado
    public static void imprimeMapa(char[][] mapa) {
        System.out.print(" 01234\n");
        for (int i = 0; i < mapa.length; i++) {
            System.out.print(i);
            for (int k = 0; k < 5; k++) {
                System.out.print(mapa[i][k]);
            }
            System.out.print('\n');
        }
    }

    // gera os mapas de cada jogador
    public static void geraMapas(char[][] mapa) {
        for (int i = 0; i < 5; i++) {
            for (int k = 0; k < 5; k++) {
                mapa[i][k] = '-';
            }
        }
    }

    // valida as coordenadas informadas pelo usuário
    public static boolean validaCoordenadas(int x, int y) {
        if ((x > 4 || x < 0) || (y > 4 || y < 0)) {
            System.out.println("Invalid coordinates. Choose different coordinates.");
            return false;
        }

        return true;
    }
}

File: LISTA 2/batalhaNaval.py
# Nome: Pedro Elias Santiago Mattar

# solicita que o jogador informe as coordenadas nas quais os navios serão colocados
def solicitaCoordenadas(jogador, coordenadas):
    entradaValida = False
    coordenadaAnteriorValida = False

    print("PLAYER " + str(jogador) + ", ENTER YOUR SHIPS COORDINATES.")

    for i in range(5):
        entradaValida = False
        coordenadaAnteriorValida = False

        while (not entradaValida or not coordenadaAnteriorValida):
            print("Enter ship " + str((i+1)) + " location: ")

            x, y = map(int, input().split())
            entradaValida = validaCoordenadas(
                x, y)

            if (entradaValida):
                coordenadas.append([x, y])

            if (i > 0):
                coordenadaAnteriorValida = validaCoordenadasAnteriores(
                    x, y, coordenadas, i)
            else:
                coordenadaAnteriorValida = True

# Valida as coordenada informada pelo usuário com as coordenadas informadas previamente
def validaCoordenadasAnteriores(x, y, coordenadas, numNavios):
    for i in range(numNavios):
        if (x == coordenadas[i][0] and y == coordenadas[i][1]):
            print("You already have a ship there. Choose different coordinates.")
            return False

    return True


# função que executa a lógica de um turno de um jogador
def turnoJogador(jogador, mapaJogador, coordenadasJogador):
    entradaValida = False
    localJaAtacado = False
    x = 0
    y = 0

    while (not entradaValida or localJaAtacado):
        print("Player " + str(jogador) + " , enter hit row/column: ")
        x, y = map(int, input().split())
        entradaValida = validaCoordenadas(x, y)

        if (entradaValida):
            localJaAtacado = verificaLocaisAtacados(x, y, mapaJogador)
            if (localJaAtacado):
                print("You already fired on this spot. Choose different coordinates.")
            else:
                marcaLocaisMapa(x, y, mapaJogador, coordenadasJogador,
                                jogador, 2 if jogador == 1 else 1)
                imprimeMapa(mapaJogador)

    return todosNaviosAfundados(mapaJogador) == 5

# Verifica se as coordenadas inforamadas pelo usuário são válidas
def validaCoordenadas(x, y):
    if ((x > 4 or x < 0) or (y > 4 or y < 0)):
        print("Invalid coordinates. Choose different coordinates.")
        return False
    return True


# verifica se as coordenadas correspondentes já foram atacadas
def verificaLocaisAtacados(x, y, mapa):
    if (mapa[x][y] == 'X' or mapa[x][y] == 'O'):
        return True

    return False

# marca locais de ataque (X) ou locais de erro (O) no mapa do jogador que está sendo atacado
def marcaLocaisMapa(x, y, mapa, coordenadas, jogadorAtaque, jogadorDefesa):
    acertou = False

    for i in range(len(coordenadas)):
        if x == coordenadas[i][0] and y == coordenadas[i][1]:
            mapa[x][y] = 'X'
            print("PLAYER " + str(jogadorAtaque) +
                  " HIT PLAYER " + str(jogadorDefesa) + "'s SHIP!")
            acertou = True
            break
    if (not acertou and mapa[x][y] == '-'):
        mapa[x][y] = 'O'
        print("Player " + str(jogadorAtaque) + " MISSED!")

# função que verifica se houve um vencedor
def verificaVencedor(mapaJogador1, mapaJogador2):
    if (todosNaviosAfundados(mapaJogador2) == 5):
        return 1
    if (todosNaviosAfundados(mapaJogador1) == 5):
        return 2
    return 0

# verifica se todos os navios já foram afundados
def todosNaviosAfundados(mapa):
    cont = 0

    for i in range(len(mapa)):
        for j in range(len(mapa[i])):
            if (mapa[i][j] == 'X'):
                cont += 1

    return cont

# marca o local dos navios no mapa
def marcaNaviosMapa(mapa, coordenadas):
    for i in range(len(mapa)):
        x = coordenadas[i][0]
        y = coordenadas[i][1]
        mapa[x][y] = '@'

# verifica se as coordenadas informadas pelo usuário já não foram informadas previamente
def imprimeMapa(mapa):
    print(" 01234")
    for i in range(len(mapa)):
        print(i, end='')
        for j in range(len(mapa[i])):
            print(mapa[i][j], end='')
        print("\n")

# Gera os mapas do jogo
def geraMapas(matriz):
    matriz = []
    for i in range(5):
        linha = ['-'] * 5
        matriz.append(linha)
    return matriz


# Função principal
def main():
    turnoJogador1 = True
    vencedor = 0
    coordenadasJogador1 = []
    coordenadasJogador2 = []
    mapaJogador1 = []
    mapaJogador2 = []

    # gera os mapas de cada jogador
    mapaJogador1 = geraMapas(mapaJogador1)
    mapaJogador2 = geraMapas(mapaJogador2)

    print("Welcome to Battleship!")

    # Solicita e marca as coordenadas dos navios do jogador 1 no mapa
    solicitaCoordenadas(1, coordenadasJogador1)
    marcaNaviosMapa(mapaJogador1, coordenadasJogador1)
    imprimeMapa(mapaJogador1)
    
    # Solicita e marca as coordenadas dos navios do jogador 2 no mapa
    solicitaCoordenadas(2, coordenadasJogador2)
    marcaNaviosMapa(mapaJogador2, coordenadasJogador2)
    imprimeMapa(mapaJogador2)

    # Gera os mapas novamente, para ocultar a posição dos navios dos jogadores
    mapaJogador1 = geraMapas(mapaJogador1)
    mapaJogador2 = geraMapas(mapaJogador2)

    # Loop principal do jogo
    while (vencedor == 0):
        if (turnoJogador1):
            vencedor = 1 if turnoJogador(
                1, mapaJogador2, coordenadasJogador2) else 0
        else:
            vencedor = 2 if turnoJogador(
                2, mapaJogador1, coordenadasJogador1) else 0

        turnoJogador1 = not turnoJogador1

    # Informa o jogador vencedor
    if (vencedor > 0):
        print("PLAYER " + str(vencedor) +
              " WINS! YOU SUNK ALL OF YOUR OPPONENT’S SHIPS!")
        print("Final boards:")
        print("Jogador 1:")
        imprimeMapa(mapaJogador1)

        print("Jogador 2:")
        imprimeMapa(mapaJogador2)


main()
